ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -0,0 +1,229 @@
+   * @return a {@link Map} of log segment name to valid size
+   */
+  private Map<String, Long> collectValidSizeByLogSegment(long deleteReferenceTimeInMs, long expirationReferenceTimeInMs)
+      throws StoreException {
+    Set<String> deleteKeys = new HashSet<>();
+    Map<String, Long> validSizePerLogSegment = new HashMap<>();
+    for (IndexSegment indexSegment : index.getIndexSegments().descendingMap().values()) {
+      List<MessageInfo> messageInfos = getEntries(indexSegment);
+      populateLogSegmentMap(validSizePerLogSegment, deleteReferenceTimeInMs, expirationReferenceTimeInMs, messageInfos,
+          indexSegment.getLastModifiedTime() * Time.MsPerSec, indexSegment.getLogSegmentName(), deleteKeys);
+    }
+    return validSizePerLogSegment;
+  }
+
+  /**
+   * Process a record in the index and populate the valid data size per container map.
+   * @param validDataSizePerContainer the valid data size per container map
+   * @param expirationReferenceTimeInMs the reference time used to decide whether or not a record is valid
+   * @param messageInfos a list of {@link MessageInfo} to be processed
+   * @param deleteKeys a {@link Set} of collected delete keys
+   * @throws StoreException
+   */
+  private void populateContainerMap(Map<String, Map<String, Long>> validDataSizePerContainer,
+      long expirationReferenceTimeInMs, List<MessageInfo> messageInfos, Set<String> deleteKeys) throws StoreException {
+    for (MessageInfo messageInfo : messageInfos) {
+      if (messageInfo.isDeleted()) {
+        // delete record
+        deleteKeys.add(messageInfo.getStoreKey().getID());
+      } else if (!isExpired(messageInfo.getExpirationTimeInMs(), expirationReferenceTimeInMs) && !deleteKeys.contains(
+          messageInfo.getStoreKey().getID())) {
+        // put record that is not expired and not deleted
+        updateNestedMapHelper(validDataSizePerContainer, BlobStoreStats.getServiceId(messageInfo.getStoreKey()),
+            BlobStoreStats.getContainerId(messageInfo.getStoreKey()), messageInfo.getSize());
+      }
+    }
+  }
+
+  /**
+   * Process a record in the index and populate the valid size per log segment map.
+   * @param validSizePerLogSegment the valid size per log segment map
+   * @param deleteReferenceTimeInMs the reference time to be used to decide whether or not a blob is deleted
+   * @param expirationReferenceTimeInMs the reference time to be used to decide whether or not a blob is expired
+   * @param messageInfos a list {@link MessageInfo} to be processed
+   * @param lastModifiedTimeInMs the time stamp indicating when the record that is being processed was added
+   * @param logSegmentName the log segment name of the record that is being processed
+   * @param deleteKeys a {@link Set} of collected delete keys
+   * @throws StoreException
+   */
+  private void populateLogSegmentMap(Map<String, Long> validSizePerLogSegment, long deleteReferenceTimeInMs,
```

Comment:
may I know where is the code to get valid data size of a put record for a delete record in the same index segment.